packages/entity-list/src/modules/preferences/sagas.js

```diff
@@ -81,3 +87,53 @@ export function* resetPreferences() {
   yield call(listSagas.setSorting)
   yield call(listSagas.reloadData)
 }
+
+export function* displayColumnModal() {
+  const {formDefinition} = yield select(listSagas.listSelector)
+  const {columns: preferencesColumns} = yield select(preferencesSelector)
+  const formColumns = formDefinition.children
+    .find(child => child.componentType === 'table')
+    .children
+    .map(column => ({
+      ...column,
+      hidden: Object.prototype.hasOwnProperty.call(preferencesColumns, column.id)
+        ? !preferencesColumns[column.id] : column.hidden
+    }))
+
+  const answerChannel = yield call(channel)
+  yield put(notifier.modalComponent(
+    `${formDefinition.id}-column-selection`,
+    'client.entity-list.preferences.columns',
+    null,
+    ({close}) => {
+      const onOk = columns => {
+        close()
+        answerChannel.put(columns)
+      }
+
+      return <ColumnPicker columns={formColumns} onOk={onOk}/>
+    },
+    true
+  ))
+
+  yield saveColumnPreferences(answerChannel, preferencesColumns, formDefinition)
+}
+
+function* saveColumnPreferences(answerChannel, preferencesColumns, formDefinition) {
+  const selectedColumns = yield take(answerChannel)
+  yield put(setColumns({...preferencesColumns, ...selectedColumns}))
+  if (Object.entries(selectedColumns).some(([, value]) => value)) {
```

Comment:
this will not work perfectly. if you open the dialog to just remove a column it will not refresh (because nothing in selected Columns has the value true")

Reply:
Thing is, if we only remove columns we don't need to refresh the whole list. Since the list knows about the preferences it will remove the column and re-render as soon as they are updated, without refreshing. The only point where we need to refresh the whole list and reload data is when we add a column to the list, since the data to display is then not already loaded.

I guess "listActions.refresh()" is not optimally named for what it does, but I wouldn't want to rename it at this point.

Thoughts?